bridge/server/user/oauth-routes.ts

```diff
@@ -0,0 +1,162 @@
+      return res.render('error', {
+        title: 'Permission denied',
+        message: 'Forbidden',
+      });
+    }
+
+    try {
+      const tokenSet = await client.callback(redirectUri, params, {
+        code_verifier: verifiers.codeVerifier,
+        nonce: verifiers.nonce,
+        scope,
+      });
+      reduceRefreshDateBy(tokenSet, reduceRefreshDateSeconds);
+      await authenticateSession(req, tokenSet);
+      res.redirect(getRootLocation());
+    } catch (error) {
+      const err = error as errors.OPError | errors.RPError;
+      console.log(`Error while handling the redirect. Cause : ${err.message}`);
+
+      if (err.response?.statusCode === 403) {
+        const response = {
+          title: 'Permission denied',
+          message: '',
+        };
+        response.message =
+          (err.response.body as Record<string, string>).message ?? 'User is not allowed access the instance.';
+        return res.render('error', response);
+      } else {
+        return res.render('error', {
+          title: 'Internal error',
+          message: 'Error while handling the redirect. Please retry and check whether the problem exists.',
+          location: getRootLocation(),
+        });
+      }
+    }
+  });
+
+  /**
+   * Router level middleware for logout
+   */
+  router.get('/logout', async (req: Request, res: Response) => {
+    if (req.query.status) {
+      return res.render('logout', { location: getRootLocation() });
+    }
+    if (!isAuthenticated(req.session)) {
+      // Session is not authenticated, redirect to root
+      return res.json();
+    }
+
+    const hint = getLogoutHint(req) ?? '';
+    if (req.session.tokenSet.access_token && client.issuer.metadata.revocation_endpoint) {
+      client.revoke(req.session.tokenSet.access_token);
+    }
+    removeSession(req);
+
+    if (client.issuer.metadata.end_session_endpoint) {
+      const params: EndSessionData = {
+        id_token_hint: hint,
+        state: generators.state(),
+        post_logout_redirect_uri: redirectUri,
+        end_session_endpoint: client.issuer.metadata.end_session_endpoint,
+      };
+      return res.json(params);
+    } else {
+      return res.json();
+    }
+  });
+
+  return router;
+}
+
+/**
+ * Sets the expiry date x seconds before the real one
+ * @param tokenSet
+ * @param seconds
+ */
+function reduceRefreshDateBy(tokenSet: TokenSet, seconds: number): void {
+  tokenSet.expires_at = tokenSet.expires_at ? tokenSet.expires_at - seconds : undefined; // token should be refreshed x seconds earlier
+}
+
+/**
+ * Delete states created by the authentication-flow if expiry is reached
+ */
+function validateStateTimes(): void {
+  for (const key of Object.keys(codeVerifiers)) {
+    if (new Date().getTime() > codeVerifiers[key]?.expiresAt) {
+      delete codeVerifiers[key];
+    }
+  }
+}
+
+if (process.env.NODE_ENV !== 'test') {
+  setInterval(validateStateTimes, 5_000); // could lead to missing exit of jest without using fakeTimers
```

Comment:
This will be changed in another PR (in progress). The session and the validation data like `state`, `nonce`, and `codeVerifier` will be saved in MongoDB. Here we also have a TTL index

Reply:
Oh, yes, you told us about that.